### PR TITLE
Scaled nodes changes

### PR DESCRIPTION
**Pull request by obino**
This PR originated from relatively rare random issues we could see during autoscaling tests. Here are the list of things done:

- limited the time and actions done in set_parameters, ensure we set both @options and @nodes only then the check are fully completed, and pushed the further action/check to enforce_options;

- ensure that no other calls can be fulfilled if @nodes is empty: this prevent bogus/spurious errors in the logs;

- prevent temporary IAAS options are accidentally propagated to the AC;

- limit locks around initialize_nodes_in_parallel.

### CODE REVIEW

**Comment by obino** [comment]
Ran the autoscale tests few times, all passed. OCD build http://ocd.appscale.net:8080/job/Daily%20Build/6373/

**Comment by jeanleonov** [comment]
13 commits.. For sure it wasn't a trivial task to debug and fix this mysterious issue!

File: AppController/djinn.rb [hunk 1]
```diff
@@ -893,81 +925,52 @@ def set_parameters(layout, options, secret)
...
+    # in the started nodes.
     @state_change_lock.synchronize {
-      ENV['EC2_URL'] = @options['ec2_url']
-      if @options['ec2_access_key'].nil?
-        @options['ec2_access_key'] = @options['EC2_ACCESS_KEY']
-        @options['ec2_secret_key'] = @options['EC2_SECRET_KEY']
-        @options['ec2_url'] = @options['EC2_URL']
+      begin
+        @nodes = check_layout(layout, checked_opts['keyname'])
+      rescue AppScaleException => e
```
Comment (cdonati) [inline AppController/djinn.rb]: Why does this exception need to be caught?

File: AppController/djinn.rb [hunk 2]
```diff
@@ -893,81 +925,52 @@ def set_parameters(layout, options, secret)
...
+        return e.message
       end
+      find_me_in_locations
+      return "Error: Couldn't find me in the node map" if @my_index.nil?
+
+      # Now we can unlock the main thread and let it proceed with the
+      # initialization.
+      @options = checked_opts
     }
+    Djinn.log_info("Successfully recevied nodes layout (#{@nodes}) and deployment options (#{@options}).")
```
Comment (cdonati) [inline AppController/djinn.rb]: received

File: AppController/djinn.rb [hunk 3]
```diff
@@ -862,7 +864,37 @@ def enforce_options
     # Set the proper log level.
     new_level = Logger::INFO
     new_level = Logger::DEBUG if @options['verbose'].downcase == "true"
-    @@log.level = new_level if @@log.level != new_level
+    @state_change_lock.synchronize {
+      @@log.level = new_level if @@log.level != new_level
+    }
+
+    # The master node can not enforce some sanity checks on the options.
```
[the remark this answers is not in the document]
Comment (obino) [reply AppController/djinn.rb]: ops thanks! That was can now ...

File: AppController/djinn.rb [hunk 4]
```diff
@@ -862,7 +864,37 @@ def enforce_options
...
+        end
+        if Integer(@options['max_machines']) < Integer(@options['min_machines'])
+          Djinn.log_warn("min_machines is bigger than max_machines!")
+          @options['max_machines'] = @options['min_machines']
+        end
+
+        # Ensure we have the correct EC2 credentials available.
+        ENV['EC2_URL'] = @options['ec2_url']
+        if @options['ec2_access_key'].nil?
+          @options['ec2_access_key'] = @options['EC2_ACCESS_KEY']
```
[the remark this answers is not in the document]
Comment (obino) [reply AppController/djinn.rb]: Not sure TBH, this code was simply relocated. It's on my TODO to check why we do that and only for AWS, but haven't quite understood it yet ...

**Comment by obino** [comment]
OCD build http://ocd.appscale.net:8080/job/Daily%20Build/6412/
multiple autoscale tests were all blue http://ocd.appscale.net:8080/job/Autoscale-test-manual-triggers/2639/

**Pull request merged**